packages/hierarchies/src/hierarchies/HierarchyProvider.ts

```diff
@@ -396,6 +426,104 @@ export class HierarchyProvider {
+      from(this.hierarchyDefinition.defineHierarchyLevel({ parentNode })).pipe(mergeAll(), shareReplay()),
+      HierarchyNodesDefinition.isCustomNode,
+    );
+
+    // query instance keys and a flag whether they should be hidden
+    const instanceKeys = instanceDefs.pipe(
+      mergeMap((def) =>
+        this._queryScheduler.scheduleSubscription(
+          of(def.query).pipe(
+            mergeMap(async (query) => {
+              const ecsql = `
+              SELECT
+                ${NodeSelectClauseColumnNames.FullClassName},
+                ${NodeSelectClauseColumnNames.ECInstanceId},
+                ${NodeSelectClauseColumnNames.HideNodeInHierarchy}
+              FROM (
+                ${query.ecsql}
+              )
+            `;
+              const reader = this.queryExecutor.createQueryReader({ ...query, ecsql }, { rowFormat: "Indexes", limit: "unbounded" });
+              return from(reader).pipe(
+                map((row) => ({
+                  key: {
+                    className: normalizeFullClassName(row[0]),
+                    id: row[1],
+                  },
+                  hide: !!row[2],
+                })),
+              );
+            }),
+            mergeAll(),
+          ),
+        ),
+      ),
+      shareReplay(),
+    );
+    // split the instance keys observable based on whether they should be hidden or not
+    const [visibleNodeInstanceKeys, hiddenNodeInstanceKeys] = partition(instanceKeys, ({ hide }) => !hide);
+
+    // hidden items' handling:
+    // - if a custom node is hidden, we'll want to retrieve instance keys of its children, otherwise we don't care about it,
+    // - if an instance key is hidden, we want to create a merged node similar to what we do in `createHideNodesInHierarchyOperator`
+    const hiddenParentNodes = merge(
+      customDefs.pipe(
+        mergeMap((def) => (def.node.processingParams?.hideInHierarchy ? this.createInitializedNodesObservable(of(def.node), parentNode) : EMPTY)),
+      ),
+      hiddenNodeInstanceKeys.pipe(
+        // first merge all keys by class
+        reduceToMergeMapList(
+          ({ key }) => key.className,
+          ({ key }) => key.id,
+        ),
+        // then, for each class, create a temp node
+        mergeMap(
+          (mergedMap): Array<ParentHierarchyNode & { key: InstancesNodeKey }> =>
+            [...mergedMap.entries()].map(([className, ids]) => ({
+              key: {
+                type: "instances",
+                instanceKeys: ids.map((id) => ({ className, id })),
+              },
+              parentKeys: [],
+              label: "",
+            })),
+        ),
+      ),
+    );
+
+    // merge visible instance keys from this level & the ones we get recursively requesting from deeper levels
+    const visibleInstanceKeys = merge(
+      visibleNodeInstanceKeys.pipe(map(({ key }) => key)),
+      hiddenParentNodes.pipe(mergeMap((hiddenNode) => this.getNodeInstanceKeys({ parentNode: hiddenNode }))),
```

Comment:
Shouldn't this be `from(this.getNodeInstanceKeys({ parentNode: hiddenNode }))`? It looks like there could be a private `getNodeInstanceKeysObs` functions that returns `Observable` instead of iterator to avoid converting iterator -> observable -> iterator?

Reply:
`mergeMap` works with `ObservableInput` and `AsyncIterableIterator` is one, so that was fine. However, I used your suggestion to extract creating observable into a separate function, so now it gets an `Observable` rather than `AsyncIterableIterator`.